src/typechecker.erl

```diff
@@ -2890,6 +2906,106 @@ check_clause(_Env, _ArgsTy, _ResTy, Term) ->
     io:format("DEBUG: check_clause term: ~p~n", [Term]),
     throw(check_clause).
 
+%% Refine types by matching clause. MatchedTys are the types exhausted by
+%% each pattern in the previous clause.
+refine_clause_arg_tys(Tys, MatchedTys, [], TEnv) ->
+    Ty        = type(tuple, Tys),
+    MatchedTy = type(tuple, MatchedTys),
+    try refine(Ty, MatchedTy, TEnv) of
+        ?type(tuple, RefTys) ->
+            RefTys;
+        ?type(none) ->
+            Tys; %% We could return [none(), ...] here if we'd want to do
+                 %% exhaustiveness checking
+        ?type(union, _) ->
+            Tys %% Multiple possibilities => don't refine
+    catch
+        no_refinement -> Tys
+    end;
+refine_clause_arg_tys(Tys, _TysBounds, _Guards, _TEnv) ->
+    Tys.
+
+%% Normalize, refine, revert normalize if no refinement.
+%% May throw no_refinement.
+refine(OrigTy, Ty, TEnv) ->
+    NormTy = normalize(OrigTy, TEnv),
+    case refine_ty(NormTy, normalize(Ty, TEnv), TEnv) of
+        NormTy -> OrigTy;
+        RefTy  -> RefTy
+    end.
+
+%% May throw no_refinement.
+refine_ty(?type(union, Tys), Ty, TEnv) ->
+    normalize(type(union, [refine(U, Ty, TEnv) || U <- Tys]), TEnv);
+refine_ty(?type(tuple, Tys1), ?type(tuple, Tys2), TEnv)
+  when length(Tys1) > 0 ->
+    %% Non-empty tuple
+    RefTys = [refine(Ty1, Ty2, TEnv) || {Ty1, Ty2} <- lists:zip(Tys1, Tys2)],
+    %% {a|b, a|b} \ {a,a} => {a|b, b} | {b, a|b}
+    PossibleTyss = [case RefTy of
+                        Ty          -> [Ty];
+                        ?type(none) -> [Ty];
+                        _           -> [Ty, RefTy]
+                    end || {Ty, RefTy} <- lists:zip(Tys1, RefTys)],
+    [_|TuplesElems] = pick_from_each(PossibleTyss),
+    Tuples = [type(tuple, TupleElems) || TupleElems <- TuplesElems],
+    normalize(type(union, Tuples), TEnv);
+refine_ty(?type(any), _Ty, _) ->
+    throw(no_refinement);
+refine_ty(Ty, ?type(none), _) ->
+    Ty;
+refine_ty({ann_type, _, [_, Ty1]}, Ty2, TEnv) ->
+    refine_ty(Ty1, Ty2, TEnv);
+refine_ty(Ty1, {ann_type, _, [_, Ty2]}, TEnv) ->
+    refine_ty(Ty1, Ty2, TEnv);
+refine_ty(?type(T, A), ?type(T, A), _) ->
+    type(none);
+refine_ty({atom, _, A}, {atom, _, A}, _) ->
+    type(none);
+refine_ty(?type(list, A), ?type(nil), _TEnv) ->
+    type(nonempty_list, A);
+refine_ty({Tag1, _, M} = Ty, {Tag2, _, N}, _TEnv)
+    when Tag1 == integer orelse Tag1 == char,
+         Tag2 == integer orelse Tag2 == char ->
+    if M == N -> type(none);
+       M /= N -> Ty
+    end;
+refine_ty(Ty, {Tag, _, V}, TEnv) when ?is_int_type(Ty),
+                                      Tag == integer orelse Tag == char ->
+    case Ty of
+        ?type(non_neg_integer) when V == 0 -> type(pos_integer);
+        ?type(non_neg_integer) when V < 0  -> Ty;
+        ?type(pos_integer)     when V =< 0 -> Ty;
+        ?type(neg_integer)     when V < 0  -> Ty;
```

Comment:
My suggestion would be to drop these cases completely and return `Ty` instead of `no_refinement` in the catch-all below. I see no reason to abort refinement when there's some imprecision (see comment on `any()` case above).

Reply:
It's a typo for sure Peter. Good catch. About Ulf's suggesion, I'm not sure yet.